### DISCUSSION TITLE
Roadmap for `ruff server`

### DISCUSSION THREAD

**Discussion by snowsignal** [astral-sh/ruff#10581] [General]
`ruff server` is still in an early stage of development. This discussion is both a public reference of our roadmap for the project, and a place to raise questions or make requests related to the server and its proposed features.

This is subject to change as the project evolves, and feedback is very much welcomed!

## Roadmap

- [x] Lint diagnostics show in Python files
- [x] Quick Fixes for diagnostics
- [x] Full-document formatting
- [x] Range formatting
- [x] Multi-threaded, lock-free architecture
- [x] In-memory document cache that tracks changes to unsaved files
- [x] NeoVim support
- [x] VSCode support (pre-release extension only, with limitations)
- [x] Uses `pyproject.toml`/`ruff.toml`/`.ruff.toml` for linter/formatter configuration, per workspace folder
- [x] Automatic configuration reloading when a config file is changed on disk
- [x] Jupyter Notebook document support
- [x] Source-level Code Actions / Commands for VS Code (Fix All, Organize Imports)
- [ ] Substantial test suite for features and common scenarios
- [x] Sublime Text support
- [x] Support for extension-specific configuration and settings
- [ ] Improved scheduler with support for event handling and task cancellation
- [x] Support for Helix, Lapce, Kate, and Zed
- [x] Proper configuration overrides for sub-folder configuration
- [ ] Optimized calls to Ruff (for example: cached diagnostics)
- [ ] Workspace-wide diagnostics window
- [ ] Task progress bar
- [ ] Test suite that directly emulates real editor sessions

**Comment by joknoll** [comment]
Is it planned to replace tools like pylsp or pyright? So basically make it a fully featured lsp with autocomplete, goto refs/impl, typechecking etc. or should it just be a complement to those?

**Reply by dhruvmanila** [reply]
Currently, the language server needs to be used alongside other server that provides capabilities like the ones that you've mentioned. However, we do want to implement those capabilities but that will require supporting it in Ruff core itself which is currently being worked on (the Red Knot project). This will require some time to complete.

**Comment by Moosems** [comment]
> `ruff server` supports the same feature set as `ruff-lsp`, but with superior performance and no installation required. `ruff server` was marked as stable in Ruff v0.5.3."

Will `ruff-lsp` be deprecated/sunset at any point in the future? If `ruff server` supports the full tool suite I'm not sure where the purpose of `ruff-lsp` lies, regardless of how much I or anyone else loved the tool when `ruff server` now fulfills the same job.

**Reply by dhruvmanila** [reply]
The current plan is to maintain `ruff-lsp` to provide any bug fixes. But, our plan is to deprecate it and archive the repository. I'll announce the plan about soon.

**Comment by knirch** [comment]
Unsure if this is ok or not, but it felt wrong to comment on a closed bug, or creating a new one when I spotted this thread.

`Automatic configuration reloading when a config file is changed on disk` is marked as done, but I ran across this when trying to figure out why I had to restart my server when editing pyproject.toml;

> "I'm not sure whether this is actually a problem or not because it's suppose to be handled on the client side. The client will send us request on file events (create, change, delete). So, this should work."
https://github.com/astral-sh/ruff/issues/11766 

This leaves me a bit confused. Why would the editor send toml files to the python lsp? At least for vim-lsp, it's my understanding that each vim instance starts their own servers, meaning if I open a new terminal and start poking around in pyproject.toml, firstly vim-lsp/vim-lsp-settings wouldn't talk to the ruff server that is currently reporting back to my main editor, and if I were to open pyproject.toml in that editor, there would have to be logic explicitly stating that if my server is ruff, then send pyproject.toml, which feels oddly specific for a client to deal with no?

I'm all for being completely wrong, but .. it does sound a lot like "vscode does this, so we can skip monitoring for changes to the configuration file(s)"

(quick disclaimer; if anything I say comes across as combative/sarcastic, I'm truly sorry, that's not the intended tone)

**Reply by MichaReiser** [reply]
It would be best to create a new issue and share details about your setup (configuration, are you using ruff-lsp or the native ruff server...). It's otherwise difficult to help with your specific issue. 

Regarding your technical question. Ruff uses the [LSP watch files features](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_didChangeWatchedFiles) to be notified about changed files. The LSP doesn't send the `pyproject.toml` itself. Instead Ruff tells the editor that it wants to be notified about changes to `ruff.toml` and `pyproject.toml` files and Ruff rescans your entire project if any such file changed.

**Reply by dhruvmanila** [reply]
I think the issue here is https://github.com/astral-sh/ruff/issues/15013.

Currently, it seems that the `workspace_refresh` client capability is hardcoded to `true` which means that the server will always try to refresh the diagnostics (on a did change watched files notification) by sending a workspace refresh request even though the client might not support it. This is also what happens in Neovim.

@knirch Thanks for raising this!
